File: frontend/src/api/purchases.ts
/**
 * 采购管理 API 接口
 *
 * 功能模块：
 * - 获取采购申请列表（支持分页、筛选）
 * - 获取采购申请详情
 * - 创建采购申请
 * - 更新采购申请
 * - 审批采购申请（批准/拒绝）
 *
 * 筛选参数（QueryParams）：
 * - status: 按状态筛选（pending/approved/purchased/rejected）
 * - applicant_id: 按申请人筛选
 * - my: 是否仅查看当前用户的申请（布尔值）
 *
 * 工作流程：
 * 1. 用户提交采购申请（status: pending）
 * 2. 管理员审批（status: approved/rejected）
 * 3. 采购完成后标记为已采购（status: purchased）
 */

import api from './index';
import type { PurchaseRequest, PurchaseFormData, ApiResponse, PaginatedResponse, QueryParams } from '../types';

/**
 * 获取采购申请列表
 *
 * @param params - 查询参数（分页、筛选）
 * @returns 分页的采购申请列表
 *
 * 使用示例：
 * ```ts
 * // 获取我的采购申请
 * const myPurchases = await getPurchases({ my: true });
 *
 * // 获取待审批的申请（管理员）
 * const pending = await getPurchases({ status: 'pending' });
 * ```
 */
export const getPurchases = (params?: QueryParams): Promise<ApiResponse<PaginatedResponse<PurchaseRequest>>> => {
  return api.get('/purchases', { params });
};

/**
 * 获取采购申请详情
 *
 * @param id - 采购申请 ID
 * @returns 采购申请详细信息
 */
export const getPurchase = (id: number): Promise<ApiResponse<PurchaseRequest>> => {
  return api.get(`/purchases/${id}`);
};

/**
 * 创建采购申请
 *
 * @param data - 采购申请信息
 * @returns 创建的采购申请信息
 *
 * 必填字段：product_name, product_code, quantity, purpose, project_name, delivery_info
 */
export const createPurchase = (data: PurchaseFormData): Promise<ApiResponse<PurchaseRequest>> => {
  return api.post('/purchases', data);
};

/**
 * 更新采购申请
 *
 * @param id - 采购申请 ID
 * @param data - 更新的采购申请信息
 * @returns 更新后的采购申请信息
 *
 * 注意：通常只能更新待审批状态的申请
 */
export const updatePurchase = (id: number, data: Partial<PurchaseFormData>): Promise<ApiResponse<PurchaseRequest>> => {
  return api.put(`/purchases/${id}`, data);
};

/**
 * 审批采购申请
 *
 * @param id - 采购申请 ID
 * @param action - 审批操作（approve/reject）
 * @param approval_comment - 审批意见
 * @returns 更新后的采购申请信息
 *
 * 权限：仅管理员可操作
 */
export const approvePurchase = (
  id: number,
  action: 'approve' | 'reject',
  approval_comment?: string
): Promise<ApiResponse<PurchaseRequest>> => {
  return api.post(`/purchases/${id}/approve`, { action, approval_comment });
};

/**
 * 批准采购申请
 *
 * @param id - 采购申请 ID
 * @param approval_comment - 审批意见（可选）
 * @returns 更新后的采购申请信息
 *
 * 权限：仅管理员可操作
 * 效果：状态变更为 approved，记录审批人和审批时间
 */
export const approvePurchaseRequest = (id: number, approval_comment?: string): Promise<ApiResponse<PurchaseRequest>> => {
  return api.post(`/purchases/${id}/approve`, { action: 'approve', approval_comment });
};

/**
 * 拒绝采购申请
 *
 * @param id - 采购申请 ID
 * @param approval_comment - 拒绝原因（可选）
 * @returns 更新后的采购申请信息
 *
 * 权限：仅管理员可操作
 * 效果：状态变更为 rejected，记录审批人、审批时间和拒绝原因
 */
export const rejectPurchaseRequest = (id: number, approval_comment?: string): Promise<ApiResponse<PurchaseRequest>> => {
  return api.post(`/purchases/${id}/approve`, { action: 'reject', approval_comment });
};

/**
 * 采购状态枚举值
 */
export const PurchaseStatuses = {
  PENDING: 'pending',
  APPROVED: 'approved',
  PURCHASED: 'purchased',
  REJECTED: 'rejected'
} as const;

/**
 * 采购状态标签映射
 */
export const PurchaseStatusLabel: Record<string, string> = {
  pending: '待审批',
  approved: '已审批',
  purchased: '已采购',
  rejected: '已拒绝'
};

/**
 * 采购状态颜色映射（用于Tag显示）
 */
export const PurchaseStatusColor: Record<string, string> = {
  pending: 'orange',
  approved: 'green',
  purchased: 'blue',
  rejected: 'red'
};

File: backend/app/api/purchases.py
# 采购申请管理 API 路由
from flask import Blueprint, request, session, g
from datetime import datetime, date
from app.models import PurchaseRequest, Supplier
from app.utils.decorators import login_required, admin_required
from app.utils.response import success_response, error_response, paginated_response
from app.utils.validators import (
    validate_required_fields,
    validate_pagination,
    validate_quantity_field,
    validate_price_field
)


def parse_date(date_string):
    """将日期字符串解析为日期对象"""
    if not date_string:
        return None
    try:
        return datetime.strptime(date_string, '%Y-%m-%d').date()
    except (ValueError, TypeError):
        return None


def get_current_user_id():
    """从 session 获取当前用户 ID"""
    return session.get('user_id')


purchases_bp = Blueprint('purchases', __name__)


@purchases_bp.route('', methods=['GET'])
@login_required
def get_purchases():
    """获取采购申请列表（支持过滤和分页）"""
    # 获取查询参数
    status = request.args.get('status')
    applicant_id = request.args.get('applicant_id', type=int)
    my_applications = request.args.get('my', type=bool)  # 获取我的申请
    page, page_size = validate_pagination()

    # 构建查询
    query = PurchaseRequest.query

    # 应用过滤条件
    if status:
        query = query.filter_by(status=status)
    if applicant_id:
        query = query.filter_by(applicant_id=applicant_id)
    if my_applications and g.user:
        query = query.filter_by(applicant_id=g.user.id)

    # 按创建时间降序排列
    query = query.order_by(PurchaseRequest.created_at.desc())

    # 分页
    pagination = query.paginate(page=page, per_page=page_size, error_out=False)

    # 转换为字典
    items = [purchase.to_dict() for purchase in pagination.items]

    return paginated_response(items, pagination.total, page, page_size)


@purchases_bp.route('/<int:request_id>', methods=['GET'])
@login_required
def get_purchase(request_id):
    """获取采购申请详情"""
    purchase = PurchaseRequest.query.get(request_id)

    if not purchase:
        return error_response('采购申请不存在', error_code='NOT_FOUND', status=404)

    return success_response(data=purchase.to_dict())


@purchases_bp.route('', methods=['POST'])
@login_required
def create_purchase():
    """创建采购申请"""
    data = request.get_json()

    # 验证必填字段
    required_fields = [
        'product_name', 'product_code', 'quantity',
        'purpose', 'project_name', 'delivery_info'
    ]
    is_valid, error_msg = validate_required_fields(data, required_fields)
    if not is_valid:
        return error_response(error_msg, error_code='VALIDATION_ERROR')

    # 验证数量字段（必须为正整数）
    is_valid, error_msg, quantity = validate_quantity_field(data, 'quantity', required=True)
    if not is_valid:
        return error_response(error_msg, error_code='VALIDATION_ERROR')

    # 验证价格字段（如果提供）
    query_price = None
    if 'query_price' in data and data['query_price'] is not None:
        is_valid, error_msg, query_price = validate_price_field(data, 'query_price', required=False)
        if not is_valid:
            return error_response(error_msg, error_code='VALIDATION_ERROR')

    order_price = None
    if 'order_price' in data and data['order_price'] is not None:
        is_valid, error_msg, order_price = validate_price_field(data, 'order_price', required=False)
        if not is_valid:
            return error_response(error_msg, error_code='VALIDATION_ERROR')

    # 从 g 或 session 获取当前用户 ID
    current_user = g.get('current_user')
    user_id = current_user.id if current_user else get_current_user_id()

    # 生成序号
    from app.extensions import db
    last_purchase = PurchaseRequest.query.order_by(PurchaseRequest.id.desc()).first()
    if last_purchase and last_purchase.sequence_number:
        sequence_number = last_purchase.sequence_number + 1
    else:
        sequence_number = 1

    # 创建采购申请
    purchase = PurchaseRequest(
        sequence_number=sequence_number,
        product_name=data['product_name'],
        brand=data.get('brand'),
        product_code=data['product_code'],
        cas_number=data.get('cas_number'),
        form=data.get('form'),
        specifications=data.get('specifications'),
        quantity=quantity,
        purpose=data['purpose'],
        applicant_id=user_id,
        project_name=data['project_name'],
        application_date=date.today(),
        delivery_info=data['delivery_info'],
        query_price=query_price,
        order_price=order_price,
        order_quantity=data.get('order_quantity'),
        delivery_date=parse_date(data.get('delivery_date')),
        delivery_quantity=data.get('delivery_quantity'),
        campus=data.get('campus'),
        building=data.get('building'),
        room=data.get('room'),
        is_hazardous_platform=data.get('is_hazardous_platform', False),
        is_institute_center=data.get('is_institute_center', False),
        is_school_warehouse=data.get('is_school_warehouse', False),
        quality_issue_info=data.get('quality_issue_info'),
        remarks=data.get('remarks'),
        status=PurchaseRequest.STATUS_PENDING
    )

    try:
        db.session.add(purchase)
        db.session.commit()

        return success_response(
            message='采购申请已提交',
            data={
                'id': purchase.id,
                'sequence_number': purchase.sequence_number,
                'status': purchase.status
            },
            status=201
        )
    except Exception as e:
        from app.extensions import db
        import traceback
        db.session.rollback()
        print(f"Purchase creation error: {e}")
        print(f"Request data: {data}")
        print(traceback.format_exc())
        return error_response(f'采购申请创建失败: {str(e)}', error_code='INTERNAL_ERROR', status=500)


@purchases_bp.route('/<int:request_id>', methods=['PUT'])
@login_required
def update_purchase(request_id):
    """更新采购申请（仅申请人或管理员）"""
    purchase = PurchaseRequest.query.get(request_id)

    if not purchase:
        return error_response('采购申请不存在', error_code='NOT_FOUND', status=404)

    data = request.get_json()

    # 从 g 或 session 获取当前用户 ID
    current_user = g.get('current_user')
    user_id = current_user.id if current_user else get_current_user_id()

    # 检查权限：只有申请人或管理员可以编辑
    if purchase.applicant_id != user_id and not (current_user and current_user.is_superuser):
        return error_response('无权编辑此申请', error_code='PERMISSION_DENIED', status=403)

    # 已审批的申请不能修改
    if purchase.status != PurchaseRequest.STATUS_PENDING:
        return error_response('已处理的申请不能修改', error_code='INVALID_OPERATION')

    # 可更新的字段
    updatable_fields = [
        'product_name', 'brand', 'product_code', 'cas_number', 'form',
        'specifications', 'quantity', 'purpose', 'project_name',
        'delivery_info', 'campus', 'building', 'room',
        'is_hazardous_platform', 'is_institute_center', 'is_school_warehouse',
        'quality_issue_info', 'remarks'
    ]

    for field in updatable_fields:
        if field in data:
            setattr(purchase, field, data[field])

    # 处理日期字段
    if 'delivery_date' in data:
        purchase.delivery_date = parse_date(data['delivery_date'])

    # 处理数量字段
    if 'order_quantity' in data:
        purchase.order_quantity = data['order_quantity']
    if 'delivery_quantity' in data:
        purchase.delivery_quantity = data['delivery_quantity']

    # 处理价格字段
    for price_field in ['query_price', 'order_price']:
        if price_field in data:
            if data[price_field] is not None:
                is_valid, error_msg, price = validate_price_field(data, price_field, required=False)
                if not is_valid:
                    return error_response(error_msg, error_code='VALIDATION_ERROR')
                setattr(purchase, price_field, price)
            else:
                setattr(purchase, price_field, None)

    try:
        from app.extensions import db
        db.session.commit()

        return success_response(
            message='采购申请更新成功',
            data={
                'id': purchase.id,
                'status': purchase.status
            }
        )
    except Exception as e:
        from app.extensions import db
        import traceback
        db.session.rollback()
        print(f"Purchase update error: {e}")
        print(traceback.format_exc())
        return error_response(f'采购申请更新失败: {str(e)}', error_code='INTERNAL_ERROR', status=500)


@purchases_bp.route('/<int:request_id>/approve', methods=['POST'])
@admin_required
def approve_purchase(request_id):
    """批准或拒绝采购申请（仅管理员）"""
    purchase = PurchaseRequest.query.get(request_id)

    if not purchase:
        return error_response('采购申请不存在', error_code='NOT_FOUND', status=404)

    if purchase.status != PurchaseRequest.STATUS_PENDING:
        return error_response('采购申请已处理', error_code='INVALID_STATUS', status=400)

    data = request.get_json()
    action = data.get('action', 'approve')

    if action not in ['approve', 'reject']:
        return error_response('无效的操作', error_code='VALIDATION_ERROR')

    # 从 g 或 session 获取当前用户 ID
    current_user = g.get('current_user')
    user_id = current_user.id if current_user else get_current_user_id()

    try:
        from app.extensions import db

        if action == 'approve':
            purchase.status = PurchaseRequest.STATUS_APPROVED
            purchase.approver_id = user_id
            purchase.approval_comment = data.get('approval_comment')
            purchase.approved_at = datetime.utcnow()
            message = '采购申请已批准'
        else:
            purchase.status = PurchaseRequest.STATUS_REJECTED
            purchase.approver_id = user_id
            purchase.approval_comment = data.get('approval_comment')
            message = '采购申请已拒绝'

        db.session.commit()

        return success_response(
            message=message,
            data={
                'id': purchase.id,
                'status': purchase.status
            }
        )
    except Exception as e:
        from app.extensions import db
        import traceback
        db.session.rollback()
        print(f"Purchase approval error: {e}")
        print(traceback.format_exc())
        return error_response(f'采购申请处理失败: {str(e)}', error_code='INTERNAL_ERROR', status=500)
